### ISSUE TITLE
I cannot authenticate using the IamAuthenticator authenticator

### ISSUE THREAD

**Issue by ArturPalka** [IBM/keyprotect-java-client#19]
**Description**:
I'm not able to list keys using IamAuthenticator because of the error it returns. I've debugged and in the `TokenRequestBasedAuthenticator` class returned response is 200 OK, the problem occurs in the converter, which returns `java.lang.NoSuchFieldError: NUMBER`

**Code**:
```
    @Test
    void kp_api_should_list_keys_from_the_intance() {
        // given
        IamAuthenticator authenticator = new IamAuthenticator(System.getenv("KP_API_KEY"));
        authenticator.setURL(System.getenv("KP_IAM_END_POINT"));
        authenticator.validate();
        IbmKeyProtectApi testService = IbmKeyProtectApi.newInstance(authenticator);
        testService.setServiceUrl(System.getenv("KP_API_END_POINT"));

        GetKeysOptions getKeysOptionsModel = new GetKeysOptions.Builder()
                .bluemixInstance(System.getenv("KP_INSTANCE_ID"))
                .build();
        Response<ListKeys> response = testService.getKeys(getKeysOptionsModel).execute();

        // when
        List<KeyRepresentation> keys = response.getResult().getResources();


        // then
        org.assertj.core.api.Assertions.assertThat(keys)
                .isNotNull();
    }
```

**Error**:
```
java.lang.RuntimeException: Error while fetching access token from token service: 

	at com.ibm.cloud.sdk.core.security.TokenRequestBasedAuthenticator.getToken(TokenRequestBasedAuthenticator.java:252)
	at com.ibm.cloud.sdk.core.security.TokenRequestBasedAuthenticator.authenticate(TokenRequestBasedAuthenticator.java:194)
	at com.ibm.cloud.sdk.core.service.BaseService.setAuthentication(BaseService.java:260)
	at com.ibm.cloud.sdk.core.service.BaseService.createCall(BaseService.java:198)
	at com.ibm.cloud.sdk.core.service.BaseService.createServiceCall(BaseService.java:236)
	at com.ibm.cloud.ibm_key_protect_api.v2.IbmKeyProtectApi.getKeys(IbmKeyProtectApi.java:478)
	at
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.NoSuchFieldError: NUMBER
	at com.ibm.cloud.sdk.core.util.GsonSingleton.registerTypeAdapters(GsonSingleton.java:63)
	at com.ibm.cloud.sdk.core.util.GsonSingleton.createGson(GsonSingleton.java:45)
	at com.ibm.cloud.sdk.core.util.GsonSingleton.getGsonWithoutPrettyPrinting(GsonSingleton.java:91)
	at com.ibm.cloud.sdk.core.util.ResponseUtils.getValue(ResponseUtils.java:149)
	at com.ibm.cloud.sdk.core.util.ResponseConverterUtils$6.convert(ResponseConverterUtils.java:137)
	at com.ibm.cloud.sdk.core.security.TokenRequestBasedAuthenticator$2.run(TokenRequestBasedAuthenticator.java:315)
	at java.base/java.lang.Thread.run(Thread.java:834)
```	

**Library version**: `0.2.0`

**Comment by gian1200** [comment]
It seems to be happening also with version `0.1.0`.

In my case it happens when upgrading from Quarkus `2.6.3.Final` to `2.7.6.Final`. When using `2.6.3` or previous, it works as expected. Maybe there is a dependency that is conflicting.

@ArturPalka, did you manage to solve the issue, or pinpoint the root cause?

**Comment by gian1200** [comment]
So, I managed to pinpoint the issue. It seems that [keyprotect-java-client](https://github.com/IBM/keyprotect-java-client) is not compatible with `com.google.code.gson` `v2.8.9` or above. The last version that works is `2.8.8`.

As a workaround one can force the version by including this in the pom.xml:

```xml
<dependency>
    <groupId>com.google.code.gson</groupId>
    <artifactId>gson</artifactId>
    <version>2.8.8</version>
</dependency>
```

Sadly, according to Snyk, that version (and earlier) are vulnerable.
https://security.snyk.io/package/maven/com.google.code.gson:gson

Note: **"keyprotect-java-client"** [uses `v2.8.6`](https://github.com/IBM/keyprotect-java-client/blob/master/pom.xml#L92)

**Comment by BrunoHenriques** [comment]
> Sadly, according to Snyk, that version (and earlier) are vulnerable. [security.snyk.io/package/maven/com.google.code.gson:gson](https://security.snyk.io/package/maven/com.google.code.gson:gson)
> 
> Note: **"keyprotect-java-client"** [uses `v2.8.6`](https://github.com/IBM/keyprotect-java-client/blob/master/pom.xml?rgh-link-date=2022-08-13T18%3A45%3A55Z#L92)

Adding a reference to NIST which at this point says that [CVE-2022-25647](https://nvd.nist.gov/vuln/detail/CVE-2022-25647) is UNDERGOING REANALYSIS.

<img width="1472" alt="image" src="https://user-images.githubusercontent.com/5439120/189005985-0f33ffb9-e1d3-47ab-add3-c92610331ba3.png">

**Comment by dinesh-venkatraman-g** [comment]
@gian1200 we have upgraded `com.google.code.gson` to the latest version `2.9.1` which has no vulnerabilities. This upgrade also required upgrading another library `com.ibm.cloud.sdk-core`. 

Can you please verify this and let us know?

**Comment by gian1200** [comment]
@dinesh-venkatraman-g, it seems that it is still not available in the MVNRepository https://mvnrepository.com/artifact/com.ibm.cloud/ibm-key-protect-sdk.

However, by manually adding `com.ibm.cloud.sdk-core` version `9.17.13` to my `pom.xml` (as a workaround) the issue went away and it worked fine.

Looking forward to seeing the new `keyprotect-java-client` version publicly released at MVNRepository.

Thanks!

**Comment by gian1200** [comment]
Hi, v0.2.1 was released a few days ago. v0.2.2 was released a few hours ago. Is there an ETA for it to be publicly available at MVNRepository (https://mvnrepository.com/artifact/com.ibm.cloud/ibm-key-protect-sdk)?

Thanks!

**Comment by dinesh-venkatraman-g** [comment]
Hi @gian1200 we have published v0.2.4 with the fix. You can find it [here](https://search.maven.org/artifact/com.ibm.cloud/ibm-key-protect-sdk). You can test with this version and let us know it looks good.  
<img width="1533" alt="image" src="https://user-images.githubusercontent.com/100373742/194625404-09d24209-afa9-4c78-9ea2-0f769e4a1713.png">

**Comment by gian1200** [comment]
I just run some tests with the new version and it worked as expected!

I only got a warning, but I think it's related to my code and the GSON that the new version uses. I'll have to take a look at it. 

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.gson.internal.reflect.ReflectionHelper (file:/C:/Users/xxxxxxxx/.m2/repository/com/google/code/gson/gson/2.9.0/gson-2.9.0.jar) to field java.time.LocalDateTime.date
```

Regarding this issue I think we can successfully close it. Thank you @dinesh-venkatraman-g, for all the work to you and your team!

**Comment by dinesh-venkatraman-g** [comment]
The reported issue is resolved with `v0.2.4` based on the comment by the user above. So, closing this issue.
